hack/generator/pkg/codegen/pipeline_status_augment.go

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package codegen
+
+import (
+	"context"
+
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astmodel"
+)
+
+func augmentSpecWithStatus() PipelineStage {
+	return MakePipelineStage(
+		"augmentSpecWithStatus",
+		"Merges information from Status into Spec",
+		func(ctx context.Context, types astmodel.Types) (astmodel.Types, error) {
+
+			newTypes := make(astmodel.Types)
+			newTypes.AddTypes(types)
+
+			// build the augmenter we will use:
+			augmenter := fuseAugmenters(
+				flattenAugmenter(newTypes), // note that this can alter newTypes as a side-effect, so it must be preloaded with all the types
```

Comment:
This sounds iffy. I can understand it needing all of the types loaded for lookups, but why does the augmenter mutate types?

I read ahead and see the place where the update happens for typenames. Why is that needed, rather than just waiting until the loop below hits the definitions corresponding to the type names.

Ohhhhh, I think I get it - the loop below only handles resources, because those are the only things that have spec and status matched up between the json and swagger. Is that right?

Reply:
It's the bit about identifying "the same" type via object structure. We can't do that just by name, we can only do it while we are walking/exploring the resource types. When we identify that two types are the same we update the spec side of that type in-place.

Perhaps it would be clearer to output a set of new types instead?

Reply:
I realized a few things at once: #1 this step of doing side-effect updates on the named types is completely unnecessary, since we can merge the types and return an unnamed object which will be named later (& #2 in fact, that's why we are now getting worse names in a bunch of cases!)

I've removed the side-effects and everything still works.